# COMPLETER

Software for correcting completeness biases in metagenome-assembled genome functional inference.

## Development

### 1. GTDB reference genome fetching

The GTDB database was searched on the 31st of August 2024 for reference genomes using the following advanced search string:

`("GTDB Taxonomy" CONTAINS "Bacteria" AND "CheckM2 Completeness" > "99" AND "CheckM2 Contamination" < "1" AND "GTDB Representative of Species" IS TRUE)`

- GTDB taxonomy contains Bacteria
- CheckM2 completeness is larger than 99%
- CheckM2 contamination is lower than 1%
- Is a GTDB representative species

The search returned 27942 entries: `development/gtdb-adv-search.tsv`.
The genomes were downloaded using the shell script generated by the GTDB search engine: `development/gtdb-adv-search-genomes.sh`.

### 2. GTDB reference genome annotation

Using the python script `development/split_references.py` 28 batches of 1000 genomes were generated.

```{sh}
python development/split_references.py \
      -i development/reference_genomes.txt \
      -o development/batches
```

Using the shell script `development/annotate_genomes.sh` genomes were annotated in batches using DRAM. The shell script clones the DRAM snakemake repository (https://github.com/alberdilab/dram.git), unzips the genomes, simplifies file names, and runs annotation in a screen session.

```{sh}
module load mamba/1.3.1
sh annotate_genomes.sh 20

#Check number of genomes
li *zip | wc -l

#Merge when done
cat output/*/annotations.tsv > output/annotations_6.tsv
```

### 3. Convert annotations to P/A of KEGG orthologs

```{sh}

cat dram*/output/*/annotations.tsv > pa_test/annotations.tsv

source activate ../isms_probe_design/conda/a13d7e97672169ee46f1a2a7d56729c0_
python annotation_to_binary.py -v pa_test/keggs.txt -i pa_test/annotations.tsv -o pa_test/kegg_pa.csv
```

## References

https://onlinelibrary.wiley.com/doi/full/10.1111/geb.13185
https://www.annualreviews.org/content/journals/10.1146/annurev-ecolsys-102221-050754
https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.14198
https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.14339

https://cran.r-project.org/web/packages/corHMM/vignettes/corHMMv2.1-vignette.pdf
