#' Imputes gene annotations to incomplete genomes based on phylogenetic signal.
#'
#' @param traits Table containing presence/absence information of gene annotations (columns) across genomes (rows).
#' @param focaltree The subseted tree generated by getfocaltree based on the focal genome
#' @param focal_genome The name of the focal genome to be imputed.
#' @param span Number of neighboring genomes to be employed for calculating the imputed values.
#' @param power Power value used when assigning weights to neighbor genomes (larger values assign larger weights to close neighbors)
#' @param threshold Relative threshold to assign a presence.
#' @return A table of imputed values and imputation statistics.
#' @importFrom ape cophenetic.phylo keep.tip drop.tip
#' @examples
#' completer(genome_traits,tree)
#' @export

imputer <- function(traits,focaltree,focal_genome,span=span,power=power, threshold=threshold){
  
  focal_target <- focal_genome$target
  focal_reference <- focal_genome$reference

  # Get all tips
  all_tips <- focaltree$tip.label
  
  # Remove focal tip to retain reference tips
  reference_tips <- all_tips[all_tips != focal_reference]
  
  # Sort tips by distance distances and remove focal genome
  focal_tip_distances <- ape::cophenetic.phylo(focaltree)[focal_reference,reference_tips] %>% sort()
  
  # Get observed data
  rownames(traits) <- NULL
  observed <- traits %>% 
    # Filter reference genomes
    filter(genome == focal_target) %>% 
    column_to_rownames(var="genome") %>% 
    t() %>% 
    as.data.frame() %>% 
    rownames_to_column(var="trait") %>% 
    rename(observed=2)
  
  # Get imputed data
  imputed <- traits %>% 
    # Filter reference genomes
    filter(genome %in% reference_tips) %>% 
    # Calculate weights based on distances and power value
    mutate(weight=(1 - focal_tip_distances)^power/sum((1 - focal_tip_distances)^power)) %>%
    # Calculate weighed value per reference genome
    mutate(across(-c(genome, weight), ~ . * weight)) %>%
    # Calculate imputed values
    summarise(across(-c(genome, weight), sum)) %>%
    # Apply imputation threshold (if value > threshold, then turn 0 to 1)
    mutate(across(everything(), ~ if_else(. > threshold, 1, 0))) %>%
    t() %>%
    as.data.frame() %>% 
    rownames_to_column(var="trait") %>% 
    rename(imputed=2) %>% 
    #Add observed traits
    left_join(observed,by="trait") %>% 
    mutate(learnt=if_else(observed == 1 , 1, imputed))
  
  return(imputed)
}
